lib/puppet/type/cisco_interface_hsrp_group.rb

```diff
@@ -0,0 +1,370 @@
+      authentication_enc_type       => 'default',
+      authentication_key_type       => 'key-chain',
+      authentication_string         => 'MyKeyChain',
+      authentication_timeout        => 'default',
+      ipv4_enable                   => true,
+      ipv4_vip                      => '10.10.10.10',
+      ipv6_vip                      => 'default',
+      ipv6_autoconfig               => 'default',
+      mac_addr                      => '00:00:11:11:22:22',
+      group_name                    => 'MyHsrpGroup',
+      preempt                       => true,
+      preempt_delay_minimum         => 100,
+      preempt_delay_reload          => 200,
+      preempt_delay_sync            => 300,
+      priority                      => 50,
+      priority_forward_thresh_lower => 20,
+      priority_forward_thresh_upper => 30,
+      timers_hello                  => 500,
+      timers_hello_msec             => true,
+      timers_hold                   => 1500,
+      timers_hold_msec              => true,
+    }
+  "
+
+  ensurable
+
+  ###################
+  # Resource Naming #
+  ###################
+
+  # Parse out the title to fill in the attributes in these
+  # patterns. These attributes can be overwritten later.
+  def self.title_patterns
+    identity = ->(x) { x }
+    patterns = []
+
+    # Below pattern matches both parts of the full composite name.
+    patterns << [
+      /^(\S+) (\S+) (\S+)$/,
+      [
+        [:interface, identity],
+        [:group, identity],
+        [:iptype, identity],
+      ],
+    ]
+    patterns
+  end
+
+  # Overwrites name method. Original method simply returns self[:name],
+  # which is no longer valid or complete.
+  # Would not have failed, but just return nothing useful.
+  def name
+    "#{self[:interface]} #{self[:group]} #{self[:iptype]}"
+  end
+
+  newparam(:name) do
+    desc 'Name of cisco_interface_hsrp_group, not used, but needed for puppet'
+  end
+
+  newparam(:iptype, namevar: true) do
+    desc 'Ip Type. Valid values are ipv4 or ipv6.'
+  end # param iptype
+
+  newparam(:group, namevar: true) do
+    desc 'HSRP group ID. Valid values are integer.'
+    munge(&:to_i)
+  end # param group
+
+  newparam(:interface, namevar: true) do
+    desc 'Name of the interface instance. Valid values are string.'
+  end # param interface
+
+  ##############
+  # Attributes #
+  ##############
+
+  newproperty(:authentication_auth_type) do
+    desc 'Authentication type'
+
+    newvalues(:cleartext, :md5, :default)
+  end # property authentication_auth_type
+
+  newproperty(:authentication_compatibility) do
+    desc 'Operate in compatibility mode for MD5 type-7 authentication.
+          Valid only for key-string'
+
+    newvalues(:true, :false, :default)
+  end # property authentication_compatibility
+
+  newproperty(:authentication_enc_type) do
+    desc 'Scheme used for encrypting authentication key string.'
+
+    newvalues(:'0', :'7', :default)
```

Comment:
Now that I see this as a manifest option, I don't think '0' and '7' are very meaningful.  I would prefer we do something similar to tacacs_server where we use meaningful names that map to these values.

1. https://github.com/cisco/cisco-network-puppet-module/blob/2f1d60e358b7009bfedcf6175357603c9e115831/lib/puppet/type/cisco_tacacs_server.rb#L120
2. https://github.com/cisco/cisco-network-node-utils/blob/develop/lib/cisco_node_utils/tacacs_server.rb#L21